Rewrite CollectionUtils dedup to work with any type

rjernst opened the pull request:
CollectionUtils contains a method for sorting and deduplicating an hppc
list of byte arrays. That is a very specific type. Yet the algorithm for
deduplicating a sorted list is very simple and does not need to be
specially typed.

This commit removes the sort portion of the method, as that is already
easily available (and timsort in Java should be just fine for this
purpose, we don't need introsort), renames to unique, and makes the
method take a generic List along with a Comparator.

relates #84735

rjernst committed:
Rewrite CollectionUtils dedup to work with any type

CollectionUtils contains a method for sorting and deduplicating an hppc
list of byte arrays. That is a very specific type. Yet the algorithm for
deduplicating a sorted list is very simple and does not need to be
specially typed.

This commit removes the sort portion of the method, as that is already
easily available (and timsort in Java should be just fine for this
purpose, we don't need introsort), renames to unique, and makes the
method take a generic List along with a Comparator.

relates #84735

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

jpountz commented on server/src/main/java/org/elasticsearch/common/util/CollectionUtils.java:
```diff
@@ -47,6 +44,27 @@ public static boolean isEmpty(Object[] array) {
+        T prevValue = list.get(0);
+        for (int i = 1; i < list.size(); ++i) {
+            T nextValue = list.get(i);
+            if (cmp.compare(nextValue, prevValue) != 0 && prevNdx++ != i) {
```
Do we need to pre-increment rather than post-increment? Otherwise it looks to me like a list where all elements are unique would still overwrite all the time?

rjernst replied:
You are right, it should be pre-increment.

jpountz commented on server/src/main/java/org/elasticsearch/common/util/CollectionUtils.java:
```diff
@@ -47,6 +44,27 @@ public static boolean isEmpty(Object[] array) {
         return array == null || array.length == 0;
     }
 
+    public static <T> void unique(List<T> list, Comparator<T> cmp) {
```
Add javadocs that the list must be sorted according to the given comparator?
Also a bit of a nit-pick, but since this modifies the list in-place, I feel like naming the method after a verb would be more appropriate?

rjernst replied:
Changed to `uniquify`

jpountz commented on server/src/main/java/org/elasticsearch/common/util/CollectionUtils.java:
```diff
@@ -47,6 +44,27 @@ public static boolean isEmpty(Object[] array) {
+        if (list.size() <= 1) {
+            return;
+        }
+
```
Should we verify that the list implements random access?

jpountz reviewed (commented):
This looks much cleaner.

rjernst committed:
address comments

rjernst committed:
support non random access list

rjernst committed:
remove trailing

rjernst commented:
@jpountz I think I addressed all your comments, can you take another look?

rjernst committed:
Merge branch 'master' into hppc/binary_dedup

rjernst committed:
spotless

grcevski committed:
Strengthen elasticsearch plugin version check (#85340)

When a plugin specifies an elasticsearch version with an empty string
in their plugin-descriptor.properties file, we don't enforce the
version check. This PR ensures that we check the version string for
empty as well as null value.

rjernst committed:
fix comparison to be unsigned

jpountz commented on server/src/main/java/org/elasticsearch/common/util/CollectionUtils.java:
```diff
@@ -47,6 +45,33 @@ public static boolean isEmpty(Object[] array) {
     }
 
+    /**
+     * Eliminate duplicates from a list.
```
nit: the `@param` javadocs already mention this but I think it's better to have it in the main javadoc too

```suggestion
     * Eliminate duplicates from a sorted list in-place.
```

jpountz reviewed (approved):
Thanks this looks good to me. One thing that makes me a bit nervous is that I'm unsure of the performance impact that this change would have for users of binary fields, but it would only possibly affect binary fields that are effectively multi-valued, which is a rare combination I suspect, so probably nothing to worry about?

amitmbm committed:
Update histogram-aggregation.asciidoc (#85356)

Fix small grammatical mistake.

Closes #85355

benwtrent committed:
Muting test InternalRealmsTests > testJwtRealmDependsOnBuildType (#85408)

relates: https://github.com/elastic/elasticsearch/issues/85407

rjernst committed:
Replace JavaVersion with Runtime.Version (#85361)

Since Java 9, the JDK has provided a means of parsing Java versions and
getting the current Java version. That class obviates the need for the
JavaVersion class of Elasticsearch. This commit removes the JavaVersion
class in favor of Runtime.Version.

Note that most of the changes here simply removed logic around
versioning because this change is intended only for the master branch,
where Java 17 is required.

rjernst committed:
Remove unused bytesref sort method from collection utils (#85353)

This commit removes leftover sort methods in CollectionUtils which were
only used in tests.

rjernst commented:
> but it would only possibly affect binary fields that are effectively multi-valued, which is a rare combination I suspect, so probably nothing to worry about?

Yes that is my thinking.

rjernst committed:
Update server/src/main/java/org/elasticsearch/common/util/CollectionUtils.java

Co-authored-by: Adrien Grand <jpountz@gmail.com>

rjernst committed:
Merge branch 'master' into hppc/binary_dedup
